lib/rgeo/shapefile/reader.rb

```diff
@@ -343,11 +343,14 @@ def next
       # and yield the Reader::Record for each one.
 
       def each
-        return to_enum(:each) unless block_given?
+        return to_enum(:each) { @num_records } unless block_given?
         return unless @opened
         # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
+        current_record_index = @cur_record_index
         rewind
         yield _read_next_record while @cur_record_index < @num_records
+      ensure
+        current_record_index && @cur_record_index = current_record_index
```

Comment:
Sorry i definitely made a mistake (i think you'll see it when testing). It should be 
```suggestion
        current_record_index && seek_index(current_record_index)
```

Otherwise we're not updating the file index!


Also @keithdoggett I'm not sure what the behavior is when the `seek_index` function returns false...

Reply:
I've actually went with
```
      def each
        return to_enum(:each) { @num_records } unless block_given?
        raise "File was not open" unless @opened
        # Each needs to be idempotent, therefore we reset all the internal indexes to their original value
        current_record_index = @cur_record_index
        rewind
        begin
          yield _read_next_record while @cur_record_index < @num_records
        ensure
          seek_index(current_record_index)
        end
      end
```